src/azure-cli/azure/cli/command_modules/vm/custom.py

```diff
@@ -3532,6 +3532,35 @@ def _get_public_ip_address_allocation(value, sku):
                 role_assignment_guid = str(_gen_guid())
                 master_template.add_resource(build_msi_role_assignment(vmss_name, vmss_id, identity_role_id,
                                                                        role_assignment_guid, identity_scope, False))
+        if encryption_identity:
+            if not cmd.supported_api_version(min_api='2023-09-01', resource_type=ResourceType.MGMT_COMPUTE):
+                raise CLIError("Usage error: Encryption Identity required API version 2023-09-01 or higher."
+                               "You can set the cloud's profile to use the required API Version with:"
+                               "az cloud set --profile latest --name <cloud name>")
+
+            if 'identity' in vmss_resource and 'userAssignedIdentities' in vmss_resource['identity'] \
+                and encryption_identity.lower() in \
+                    (k.lower() for k in vmss_resource['identity']['userAssignedIdentities'].keys()):
+
+                if 'virtualMachineProfile' not in vmss_resource['properties']:
+                    vmss_resource['properties']['virtualMachineProfile'] = {}
+                if 'securityProfile' not in vmss_resource['properties']['virtualMachineProfile']:
+                    vmss_resource['properties']['virtualMachineProfile']['securityProfile'] = {}
+                if 'encryptionIdentity' not in vmss_resource['properties']['virtualMachineProfile']['securityProfile']:
+                    vmss_resource['properties']['virtualMachineProfile']['securityProfile']['encryptionIdentity'] = {}
+
+                vmss_securityProfile_EncryptionIdentity \
+                    = vmss_resource['properties']['virtualMachineProfile']['securityProfile']['encryptionIdentity']
+
+                if 'userAssignedIdentityResourceId' not in vmss_securityProfile_EncryptionIdentity or \
+                        vmss_securityProfile_EncryptionIdentity['userAssignedIdentityResourceId'] \
+                        != encryption_identity:
+                    vmss_securityProfile_EncryptionIdentity['userAssignedIdentityResourceId'] = encryption_identity
+                    vmss_resource['properties']['virtualMachineProfile']['securityProfile']['encryptionIdentity'] \
+                        = vmss_securityProfile_EncryptionIdentity
+            else:
+                raise CLIError("Encryption Identity should be an ARM Resource ID of one of the "
+                               "user assigned identities associated to the resource")
```

Comment:
Please use the specific error type `ArgumentUsageError` in stead of `CLIError`
```suggestion
                raise ArgumentUsageError("Encryption Identity should be an ARM Resource ID of one of the "
                               "user assigned identities associated to the resource")
```

Reply:
During VMSS creation if there is any exception, it will throw cliError exception by default, that's why used that cliError exception here

Reply:
In fact, we usually recommend using more specific error type if they can be clearly classified, as this will help us in future Telemetry data analysis

Reply:
In fact, we usually recommend using more specific error type if they can be clearly classified, as this will help us in future Telemetry data analysis